mautic/mautic
# Fix tokens for email_stats and add support to redirect urls for it

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
If you use `{contactfield}` token in URL, then it's not correctly  set to `email_stats` to tokens column. This PR fixed it. Also add migration to last ID of email to prevent replace it for old emails



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2.  Create email without tokens in text and add  text with href link to `www.google.com?e={contactfield=email}`
3. Add plain text just custom string (for example `xxx`)
4.  Send email to 3 contacts
```
test1@test.com
test2@test.com
test3@test.com
```
5. Check `email_stats` table `tokens` column. See the email token has the wrong data. `{contactfield=email}` set for all with same email address
6. Edit one of the contact email , for example change `test1@test.com` to `test1changed@test.com`
7. Open sent email to `test1@test.com` and click on the link 
8. You will redirect to `www.google.com?e=test1changed@test.com` If If email was sent before change
9. Apply PR and run migrations
10. Clone email above and send
11. Check `email_stats` table `tokens` column. See the email token colmun has correct data for `{contactfield=email}`
12. Change contact `test1changed@test.com` to  `test1@test.com` 
13. Open sent email to`test1changed@test.com` and click on the link 
8. You will redirect to `www.google.com?e=test1changed@test.com`  what is correct email address during the send of email
